### PR TITLE
Cordapp distribution service design

### PR DESCRIPTION
**Pull request by ivanr3**
This PR adds a design doc for CorDapp distribution service

### CODE REVIEW

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    -Dconfiguration="config.properties"
+
+```
+
+Exact configuration parameters will be documented during the implementation.
+
+##### cds-cordapp
+
+`cds-cordapp` will provide the following functionality on *top* of the `cds-lib`:
+* Scheduled state for members to periodically sync their local repos with a single remote repository.
```
Comment (mikehearn) [inline bn-apps/cordapp-updates-distribution/design/design.md]: It's probably well past time that we just allow flows to be registered as scheduled directly. It's kind of weird that you need a state to trigger recurrent flows.

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+* Flows for BNO to notify members about new CorDapp version availability. Members CorDapp will automatically attempt to download a new version.
+* Flows for BNO to notify members about CorDapp version revocations. If a version of CorDapp was revoked, members are expected to manually update their nodes to the last not-revoked version ASAP.
```
Comment (mikehearn) [inline bn-apps/cordapp-updates-distribution/design/design.md]: What does the flow really do then, if the only outcome is manual? Maybe for now we can descope revocations and just ask them to email their users and tell them to downgrade?

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+* Flows for BNO to collect reports from members about CorDapp versions installed on their nodes. Only versions of CorDapps related to *this* Business Network should be reported. CDS will rely on the information provided in CorDapp `MANIFEST` files until a better API is available.
```
Comment (mikehearn) [inline bn-apps/cordapp-updates-distribution/design/design.md]: How do we know the relationship between app and BN? Does package namespace ownership help us here?

Comment (mikehearn) [reply bn-apps/cordapp-updates-distribution/design/design.md]: If each node has its own local Maven cache, then I guess the contents of this repo is what's "installed" and we could report the coordinates directly?

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+* Flows for members to manually request lists of available / revoked versions and their descriptions from BNO.
```
Comment (mikehearn) [inline bn-apps/cordapp-updates-distribution/design/design.md]: We should make sure all these flows are shell friendly, and perhaps just add a set of shell commands that act as aliases, to give a slightly more discoverable and enjoyable shell experience.

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+* Flows for BNO to optionally notify members if they have an outdated or revoked version of CorDapp installed on their nodes.
```
Comment (mikehearn) [inline bn-apps/cordapp-updates-distribution/design/design.md]: This list sounds like notification of updates is both push *and* pull? Why do both - why not just have the node pull this info?

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+> It's important to emphasise, that in the Business Networks where members host nodes by themselves, BNO can only *do their best* to encourage members to upgrade by notifying them via CDS or sending them an email. Ultimately it will be up to a member to decide on whether they would like to upgrade or not. It should be in a member's best interest to promptly upgrade as otherwise they might loose ability to transact if a CorDapp is not backwards compatible.
```
Comment (mikehearn) [inline bn-apps/cordapp-updates-distribution/design/design.md]: typo: lose

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+> CorDapps can be designed to stop working if a newer version is available. This can be done by making flows to compare the current version against the list of available versions from CDS. If the current version is not the latest - the flows might refuse to start. It will be up to the CorDapp developers to utilise such technics.
```
Comment (mikehearn) [inline bn-apps/cordapp-updates-distribution/design/design.md]: typo: techniques

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+There are no strict limitations around following one architecture or the other. BNs can mix and match depending on their requirements. While HTTP transport might seem to be an easier option, using Corda flows have a couple of benefits over it, in particular:
+* BNO will be able to host Maven in a private segment of their network, without exposing it to internet, effectively hidden behind firewalls and Corda.
+* BNO will be able to expose the repo to their BN members only by integrating `cds-cordapp` with Business Network Management Service.
+
+#### Implementation details
+
+The proposed solution will require implementation of the following components:
+
+* Custom Maven Resolver transports over flows and RPC. This will require the following interface to be implemented, which should be fairly straightforward to do:
```
Comment (mikehearn) [inline bn-apps/cordapp-updates-distribution/design/design.md]: Yes sorta kinda. Problem is the need to annotate things with `@Suspendable` in Java 8. Once we upgrade to a newer Java and a newer quasar this goes away. Until then we might need to fork Maven Resolver and add some annotations ourselves - possibly also tweaking it to avoid checkpointing huge amounts of stuff (we can also potentially just not checkpoint at all and assume/ensure MR is idempotent)

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+* Fetch metadata about a single or a range of versions from a remote Maven repository.
+* Download a single or a range of versions from a remote repository. `cds-lib` is packaging-type agnostic and can be used to pull down any type of artifacts from any Maven2 compatible repository. Support of classifiers will allow devs to publish CorDapps targeting different hardware / software configurations.
+
+`cds-lib` will support pluggable transports:
+* *HTTP(s)*. Available in Maven Resolver out-of-the-box with proxy- and repository- level authentications support.
+* *Corda Flows*. Transport implementation that allows transferring files over Corda Flows.
+* *Corda RPC*. Its essentially the same as *Corda Flows*, with the difference that the *RPC* version will be used if `cds-lib` is invoked from outside of a Corda node.
```
[the remark this answers is not in the document]
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Makes sense, lets descope RPC for now and add it later on if there is a need.

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+`cds-lib` will be invokable from a command line, with command line parameters taking precedence over the configuration file. CLI invocation *could* look like:
```
[the remark this answers is not in the document]
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Makes sense, renamed.

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 1 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Yeah, this API looks a bit odd to me too. Would be great to support scheduling annotations, like `@Scheduled(interval=1440000)` and maybe added another plugin similarly to `SerialisationWhitelist` to schedule thirdparty flows.

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 3 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Re: package namespace ownership. Yes, I think so. At least for the BNs that publish ownership information to the zone. I'll add a couple of lines about that.

Re: Maven cache. What is in maven cache is not necessarily what is installed, no? I.e. it might be a while in between you download something and install it on your node. Maven cache will contain all versions, not just the current one.

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 4 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Yeps, sounds good. Do we have something on the backlog to support custom shell extensions? Maybe API to allow cordapps to advertise their custom shell commands? Or otherwise we'd have to work from the assumption that CDS is installed on a node.

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 5 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: I kept both to satisfy 2 scenarios - frequent and infrequent updates.I.e if an update is released once in a while, it doesn't make any sense for members to constantly pull BNO. Especially given that BNO will have to distribute memberships as well (and maybe some other data the future).

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 6 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Fixed

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 8 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Great point. Let me reflect this.

Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Is it possible to avoid checkpointing without making changes to Corda?

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 7 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Fixed

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 2 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: You mean descope revocations at all? Or just the "notification" part of it?
I agree re: email, but it would be quite convenient to have an API to query all revoked versions. Especially for the parties who have not been the members since the very beginning.

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 4 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Yeps, will a couple of lines about that

**Comment by ivanr3** [comment]
@mikehearn added fixes, please have a look

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+| Date                 | 22 Aug 2018                                                |
+| Author               | Ivan Schasny, Mike Hearn |
+| Distribution         | Design Review Board, Product Management, Solutions Engineering, Platform Delivery |
+| Corda target version | OS                                                   |
+
+## HIGH LEVEL DESIGN
+
+### Overview
+
+This proposal describes the architecture of a reference implementation for a CorDapp Distribution Service (CDS), that would allow BNOs to distribute CorDapps updates to Business Network participants. This design targets scenarios when Business Network members maintain their nodes by themselves and can decide on what version of a CorDapp they would like to upgrade to. This design does *not* target scenarios, when upgrade function is centralised and the whole network can be shut down to perform stop-upgrade-restart.
```
Comment (gendal) [inline bn-apps/cordapp-updates-distribution/design/design.md]: "CorDapp updates" or "updates to CorDapps"

Comment (gendal) [reply bn-apps/cordapp-updates-distribution/design/design.md]: does *not* target scenarios where (no comma and where, not when)

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+### Requirements
+
+The following requirements have been gathered from various internal discussions within R3 and from [groups.io mailing lists](https://groups.io/g/corda-dev/message/190?p=,,,20,0,0,0::relevance,,updates+distribution,20,2,0,22686107):
+
+* BNO should be able to notify members about new CorDapp version availability
+* BNO should be able to revoke a specific version of CorDapp from their BN
+* A member should be able to download a specific version of CorDapp
+* BNO should be able to collect a report from members about CorDapp versions installed on their nodes
+* Members should be able to subscribe to a repository channel and be aware of it's synchronicity
```
Comment (gendal) [inline bn-apps/cordapp-updates-distribution/design/design.md]: its

Comment (gendal) [reply bn-apps/cordapp-updates-distribution/design/design.md]: actually.. what does "its synchronicity" mean?

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+### Overview
+
+This proposal describes the architecture of a reference implementation for a CorDapp Distribution Service (CDS), that would allow BNOs to distribute CorDapps updates to Business Network participants. This design targets scenarios when Business Network members maintain their nodes by themselves and can decide on what version of a CorDapp they would like to upgrade to. This design does *not* target scenarios, when upgrade function is centralised and the whole network can be shut down to perform stop-upgrade-restart.
+
+### Requirements
+
+The following requirements have been gathered from various internal discussions within R3 and from [groups.io mailing lists](https://groups.io/g/corda-dev/message/190?p=,,,20,0,0,0::relevance,,updates+distribution,20,2,0,22686107):
+
+* BNO should be able to notify members about new CorDapp version availability
+* BNO should be able to revoke a specific version of CorDapp from their BN
```
Comment (gendal) [inline bn-apps/cordapp-updates-distribution/design/design.md]: what does it mean to "revoke" a CorDapp?

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+* A member should be able to download a specific version of CorDapp
```
Comment (gendal) [inline bn-apps/cordapp-updates-distribution/design/design.md]: of *a* CorDapp

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+* BNO should be able to collect a report from members about CorDapp versions installed on their nodes
+* Members should be able to subscribe to a repository channel and be aware of it's synchronicity
+* It should be possible to prevent a CorDapp from working if a newer version is available
```
Comment (gendal) [inline bn-apps/cordapp-updates-distribution/design/design.md]: what does this mean in practical terms?  That it must be possible to stop somebody's node (which they control) from running a particular piece of code? (How?).  That it must be possible to prevent transactions created by a particular version from being accepted by other nodes? something else?

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+* CDS should support release channels
```
Comment (gendal) [inline bn-apps/cordapp-updates-distribution/design/design.md]: what does this mean?

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+* CDS should be integratable into CI/CD pipelines
+* BNO should be able to provide a textual description along with each CorDapp version
+* Members should be aware of current and revoked versions of CorDapps
+
+### Non-requirements:
+
+* CDS will not provide functionality for automatic updates installation. Node administrator will still have to stop-upgrade-restart their nodes manually.
+* CDS is not intended to be used to update the platform itself.
+* How to design a CorDapp that supports upgrades is out of scope of this design document. Information about flow versioning, states evolution and contract constraints can be found in the [Corda Docs](https://docs.corda.net)
```
Comment (gendal) [inline bn-apps/cordapp-updates-distribution/design/design.md]: *where* in the docs?  Pointing to the front page is a bit of a cop out :)

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 11 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Thanks, fixed

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 12 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Fuzz raised this originally. It means that participants should be able to sync their local repositories with the remote.  Rephrased to make it clear.

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 16 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: This was originally raised by Fuzz and also mentioned quite a lot in the discussion on groups.io. Users wanted to have something akin to Chrome release channels, i.e. to be able to distribute alfa / beta / release / etc. versions of a CorDapp to their networks. I.e. a user should be able to "subscribe" to a "beta" channel and install "beta" versions of a CorDapp to his node if he wishes so.  I expanded this further down in the design 

```
CDS won't natively support *release channels*. Release channels can be simulated by using multiple Maven repositories (a repo per channel). Users can sync down different remote repositories to different local locations and then manually install a desired version of CorDapp to their nodes.
```

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 17 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Makes sense, added more information about it

**Comment by ivanr3** [comment]
@gendal Thanks for your review. All comments definitely make sense. To avoid duplication, I wrote this design under the assumption that readers would quickly skim through the groups.io mailing list here: https://groups.io/g/corda-dev/topic/22686107#190.  I added an explicit note that the mailing list is recommended for reading prior to the design.

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 13 above)
Comment (gendal) [reply bn-apps/cordapp-updates-distribution/design/design.md]: thanks for the clarification.  In which case, I think we should reword the requirement...  to something like: "BNO should be able to notify members that they a specific version of a CorDapp should no longer be considered valid, and members should be able to configure their nodes to use such notifications to prevent such applications from executing" (is that the right wording though?  ie is this about stopping a node from running the app?  or about having an node know it should reject transactions that reference a particular version of the code?)

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 15 above)
Comment (gendal) [reply bn-apps/cordapp-updates-distribution/design/design.md]: ok - perhaps reword this too?  I think we need to be really clear about when the platform is able to enforce something versus when the platform/CDS code is able to provide the capability for somebody to write an app that enforces something.  ie I think we need to avoid writing things that lead people into incorrect mental models about how computers work :)

Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Rephrased a bit. Hopefully makes more sense now.

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 13 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Makes sense. Added clarification about what revocation in this context actually means.

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 18]
```diff
@@ -55,62 +55,53 @@ The proposal is to distribute CorDapps via Maven repositories. This would allow
...
-> java -jar cds-lib.jar
-    -Dcordapp="com.my.company.name:corda-name:1.0"
-    -DremoteRepo="https://my-company-repo.com"
-    -DlocalRepo="/path/to/my/local/repo"
-    -Dtransport="rpc"
-    -Dconfiguration="config.properties"
-
 # Will pull down all versions of cordapps and their dependencies
-> java -jar cds-lib.jar
+> java -jar corda-updates-core.jar
```
Comment (mikehearn) [inline bn-apps/cordapp-updates-distribution/design/design.md]: We have a standard way to do CLIs now, see Tommy's talk from today or ask Anthony.

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+
+### Overview
+
+This proposal describes the architecture of a reference implementation for a CorDapp Distribution Service (CDS), that would allow BNOs to distribute updates to CorDapps to Business Network participants. This design targets scenarios when Business Network members maintain their nodes by themselves and can decide on what version of a CorDapp they would like to upgrade to. This design does *not* target scenarios where upgrade function is centralised and the whole network can be shut down to perform stop-upgrade-restart.
+
+### Requirements
+
+The following requirements have been gathered from various internal discussions within R3 and from [groups.io mailing lists](https://groups.io/g/corda-dev/topic/22686107#190), which is recommended for reading prior to the design.
+
+* BNO should be able to notify members about new CorDapp version availability.
```
Comment (tudor-malene) [inline bn-apps/cordapp-updates-distribution/design/design.md]: Just seeing this design now.

Qq: When talking about "CorDapp", I assume this means both the contract code and the flows code, which live in separate jars, and will have a different lifecycle?

File: bn-apps/cordapp-updates-distribution/design/design.md [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+* BNO should be able to revoke a specific version of CorDapp from their BN. In this context revocation means that BNO should be able to notify members that a specific version of a CorDapp should no longer be considered valid, so members can promptly update their nodes.
```
Comment (tudor-malene) [inline bn-apps/cordapp-updates-distribution/design/design.md]: Does this apply to both jars?
For the contract jar, it's a bit more tricky as that is hardcoded into existing transactions (via hash).

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 19 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Yeah, great point @tudor-malene . I guess the focus of this design is on _distribution_ of the apps. I.e. how to physically get jar files to an end-user's computer. That's why it explicitly says in the beginning that this design doesn't aim on automating _installation_ of the cordapps. Node admins would still have to stop / upgrade / restart etc.

File: bn-apps/cordapp-updates-distribution/design/design.md (hunk 20 above)
Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: BNO can't forcefully revoke any jar. He can just try to encourage people to update by letting them know which jars are potentially malicious.

Currently we left the revocation out-of-scope of the first implementation. Btw this is the PR if you are interested: https://github.com/corda/corda-solutions/pull/41

Comment (tudor-malene) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Hey @ivanr3, I tried to look at it but it's a bit overwhelming. 

I'm just on a general crusade against the term "CorDapp" because it means 2 different things, and it's confusing.

Comment (ivanr3) [reply bn-apps/cordapp-updates-distribution/design/design.md]: Agree. That is why we tried to keep the distribution aside from installation to not to open a can of worms :)
